T2775

mjschranz opened the pull request:
https://webmademovies.lighthouseapp.com/projects/65733/tickets/2775-improve-autosave-experience

mjschranz committed:
[#2775] New projects will start backup when loading from dashboard, but bail early if no changes are made

mjschranz committed:
Convert system to use push state and be dialog less.
If an auto save is present that matches the project id in the URL, use it and auto save.
If an auto save is present and doesn't match the project in the URL, load the url project and notify user in header (currently fugly).
If neither are met, load project in url/default

mjschranz committed:
Purge repo of old backup dialog stuffs

mjschranz committed:
Header message still fugly, but implementation is a little better

mjschranz committed:
Remove temporary text

mjschranz committed:
Lint fixes

mjschranz committed:
Reworked logic to be cleaner

secretrobotron commented on src/ui/header.js:
```diff
@@ -117,6 +123,23 @@ define([ "dialog/dialog", "util/lang", "text!layouts/header.html", "ui/user-data
+
+          _projectAutoSave.classList.remove( "hidden" );
+          _projectAutoSave.href = queryString + "?loadAutoSave=true";
+          _projectAutoSave.innerHTML = "Hey! We found an auto save for <span class='autosave-name'>" + project.name + "</span>, last saved " +
```
You can probably place the majority of this in the layout, and only manipulate the `innerHTML` of the `autosave-name` span, right?

secretrobotron commented on src/core/project.js:
```diff
@@ -289,6 +302,8 @@ define( [ 'core/eventmanager', 'core/media' ],
     // to be saved (i.e., it has been changed since last save, or was never
     // saved before).
     _this.save = function( callback ) {
+      var history = window.history;
```
Probably doesn't need to be cached. You can just reference `window.history` directly below, right?

mjschranz committed:
Refactor project loading logic to happen inside Project module

mjschranz committed:
Fix bugs with when auto project was being used in header and also if auto save was selected when there was no project used

mjschranz committed:
Refix auto loading. Somehow it broke? Here's to hoping this sticks

mjschranz committed:
OKAY FOR SERIOUS THIS TIME

secretrobotron commented on src/core/project.js:
```diff
@@ -246,9 +241,22 @@ define( [ 'core/eventmanager', 'core/media' ],
+
+        // This means the project we imported as a back up as well for an existing project
+        // However, if the recovered auto save doesn't match the id of the project they loaded
+        // we shouldn't auto save the backup
```
Missing period.

secretrobotron commented on src/core/project.js:
```diff
@@ -246,9 +241,22 @@ define( [ 'core/eventmanager', 'core/media' ],
+        _id = json.projectID;
+
+        // This means the project we imported as a back up as well for an existing project
+        // However, if the recovered auto save doesn't match the id of the project they loaded
```
s/they/"the user"

secretrobotron commented on src/core/project.js:
```diff
@@ -344,35 +365,58 @@ define( [ 'core/eventmanager', 'core/media' ],
+          location = window.location.search,
+          loadAutoSave = location.indexOf( "loadAutoSave" ) > -1 ? true : false;
+
+      if ( !fallbackCB ) {
```
For brevity's sake, you can probably just use `fallbackCB = fallbackCB || function() {};`.

secretrobotron commented on src/core/project.js:
```diff
@@ -344,35 +365,58 @@ define( [ 'core/eventmanager', 'core/media' ],
+      }
+
+      if ( !readyCB ) {
+        readyCB = function() {};
```
Same as https://github.com/mozilla/butter/pull/1316/files#L1R377

secretrobotron commented on src/ui/header.js:
```diff
@@ -1,5 +1,7 @@
+define(
+  [ "dialog/dialog", "util/lang", "text!layouts/header.html", "ui/user-data", "ui/webmakernav/webmakernav",
+    "ui/widget/tooltip", "util/time" ],
+  function( Dialog, Lang, HEADER_TEMPLATE, UserData, WebmakerBar, ToolTip, TimeUtil ) {
```
I've been trying to keep the newlines consistent in the function signature with the newlines in the array above it so that you can easily see which objects belong to which source files. Mind cutting this line the way you did the one on top of it?

secretrobotron commented on src/ui/header.js:
```diff
@@ -101,6 +104,9 @@ define([ "dialog/dialog", "util/lang", "text!layouts/header.html", "ui/user-data
         toggleSaveButton( true );
         toggleShareButton( true );
+
+        butter.project.autosave = _projectAutoSave.onclick = null;
```
I'm looking at this again, and I think pushing logic for `butter.project.autosave` back into `project` altogether is a good idea. I don't think `header` code should be telling `butter.project.autosave` whether it's `null`, or otherwise.

secretrobotron commented on src/ui/header.js:
```diff
@@ -117,6 +123,33 @@ define([ "dialog/dialog", "util/lang", "text!layouts/header.html", "ui/user-data
+            _projectAutoSave.onclick = function() {
+              // Rebackup the project with a flag attached for loading it
+              butter.project.backupData( project );
+              butter.project.autosave = null;
```
Same as https://github.com/mozilla/butter/pull/1316/files#L8R108

secretrobotron commented on src/ui/header.js:
```diff
@@ -117,6 +123,33 @@ define([ "dialog/dialog", "util/lang", "text!layouts/header.html", "ui/user-data
+
+            _projectAutoSave.classList.remove( "hidden" );
+            _projectAutoSave.href = queryString + "?savedDataUrl=/api/project/" + project.projectID;
+            project.isAutoSave = true;
```
Same as https://github.com/mozilla/butter/pull/1316/files#L8R108.

mjschranz committed:
Header shouldn't call logout to set default state. Just make that the default state in the HTML/css.
Prevent us from pushing to the history multiple times when we only need to do it once.
Ensure project id is removed from state and toggled back in when logged in properly

mjschranz committed:
Added an appropriate comment

secretrobotron commented on src/core/project.js:
```diff
@@ -149,6 +156,16 @@ define( [ 'core/eventmanager', 'core/media' ],
         enumerable: true
+      },
+
+      "projectHistoryState": {
```
This probably doesn't need to be a property defined like this. Anything wrong with just having `this.projectHistoryState`?

secretrobotron commented on src/ui/header.js:
```diff
@@ -265,8 +311,8 @@ define([ "dialog/dialog", "util/lang", "text!layouts/header.html", "ui/user-data
-    //Default view
-    _this.views.logout();
+    // Default state is the save button should be active
+    _saveButton.addEventListener( "click", saveProject, false );
```
Can you use `toggleSaveButton( true );` instead?

secretrobotron commented on src/core/project.js:
```diff
@@ -359,35 +399,54 @@ define( [ 'core/eventmanager', 'core/media' ],
+        projectBackup = JSON.parse( projectBackup );
+      } catch( e ) { }
+
+      if ( projectBackup && projectBackup.isAutoSave ) {
```
The check for `projectBackup` can just happen once here (if you wrap this in another if block) removing the need for the `else if` below (just becomes an `else`).

secretrobotron commented on src/core/project.js:
```diff
@@ -359,35 +399,54 @@ define( [ 'core/eventmanager', 'core/media' ],
+    // Check for an existing project that was autosaved but not saved.
+    // Returns project backup data as JS object if found, otherwise null.
+    _this.checkForBackup = function( readyCB, fallbackCB ) {
+      // See if we already have a project autosaved from another session.
```
This comment looks out of place here. Should it be above the `try` below?

secretrobotron commented on src/core/project.js:
```diff
@@ -359,35 +399,54 @@ define( [ 'core/eventmanager', 'core/media' ],
+      try {
+        projectBackup = __butterStorage.getItem( "butter-backup-project" );
+        projectBackup = JSON.parse( projectBackup );
+      } catch( e ) { }
```
A warning should happen in this catch. It's a bizarre enough situation if a backup fails like this that we'd want to know.

mjschranz committed:
Lint errors.
Also made sure autosave notification went away from present but then user logged out

humphd commented on src/core/project.js:
```diff
@@ -359,35 +386,59 @@ define( [ 'core/eventmanager', 'core/media' ],
-    }
+    // Check for an existing project that was autosaved but not saved.
+    // Returns project backup data as JS object if found, otherwise null.
+    _this.checkForBackup = function( readyCB, fallbackCB ) {
```
We use 'callback' vs. 'CB' everywhere else in our tree, do it here too.

humphd commented on src/core/project.js:
```diff
@@ -359,35 +386,59 @@ define( [ 'core/eventmanager', 'core/media' ],
+          _this.import( projectBackup );
+          readyCB( _this );
+        } else {
+          var id = location.substring( location.lastIndexOf( "/" ) + 1 );
```
Use util/uri.js to parse this, in case we add more things to the URL later.  Also, after you get the `savedDataUrl` query key, use a regex to clearly show what you're meaning to get (in case our URL structure changes, this will be more clear what you meant by the last /):

```
/\/api\/project\/(\d+)/
```

humphd commented on src/default-config.json:
```diff
@@ -5,7 +5,7 @@
   "snapshotHTMLOnReady": false,
   "scrapePage": true,
-  "backupInterval": 30000,
+  "backupInterval": 10000,
```
Make sure you bump this back to 30

humphd commented on src/main.js:
```diff
@@ -784,43 +784,21 @@
             preparePage( function(){
               moduleCollection.ready( function(){
+
+                function readyCB( project ) {
```
Fix naming here too

mjschranz committed:
Dont' save a project loaded from backup is project ids match. notify user that it was loaded from backup to let them know they need to save it to keep changes

mjschranz committed:
Use a URI parsers plus regex to prevent against future api changes

mjschranz committed:
More safeguards. More callbacks

mjschranz committed:
Hi

humphd commented on src/core/project.js:
```diff
@@ -271,18 +278,27 @@ define( [ 'core/eventmanager', 'core/media' ],
+      var data;
+
       // If the project isn't different from last time, or if it's known
       // to not fit in storage, don't bother trying.
```
This comment is no longer in sync with the code it precedes, please update to reflect what's really happening.

humphd commented on src/core/project.js:
```diff
@@ -271,18 +278,27 @@ define( [ 'core/eventmanager', 'core/media' ],
+        data.backupDate = Date.now();
+      } else {
+        data = backup;
+        data.loadMe = true;
```
This name is not great, can you do better?  I'm still confused by it even after all your explanations.  It shouldn't be named in a way that relates to UI, but to a Project and the logic of a Project.

humphd commented on src/core/project.js:
```diff
@@ -359,35 +382,45 @@ define( [ 'core/eventmanager', 'core/media' ],
+          readyCallback( _this );
+          return;
+        } else {
+          _this.autosave = projectBackup;
```
Don't use "autosave"

humphd commented on src/layouts/header.html:
```diff
@@ -5,21 +5,29 @@
+      <li>
+        <div class="butter-project-autosave">
+          <a class="backup-available hidden" target="_blank">
+            Hey! We found an auto save for <span class="backup-name"></span>
```
"We found a backup for {title} from {time} ago."
